### DISCUSSION TITLE
Dispenser Improvements (OWNED / BURN address types)

### DISCUSSION THREAD

**Discussion by jdogresorg** [CounterpartyXCP/CIPs#103] [CIPs]
# Current Dispenser Address Types
Currently dispensers are able to be opened on 2 different address types :
1. `SOURCE` address
2. `EMPTY` address

An `EMPTY` address is defined in CP as any address which does not have any CP asset balance history associated with it. This means that anyone can open a dispenser up on any address, even one that they do not own!

This `EMPTY` address functionality has some pros (xcp-20 dispensers, can fund/open dispenser in a single tx), and cons (dispensers on hot-wallet addresses, can't close/refill from SOURCE address)

# Proposed Dispenser Address Types
## 1. `SOURCE` address
`SOURCE` addresses would continue to operate the same as they do now.

## 2. `OWNED` address
`OWNED` addresses would be a new address type in Counterparty, which would allow one to OPEN, REFILL, and CLOSE a dispenser on an `OWNED` address. `OWNED` addresses would need to verify ownership of the address via some method (probably signed message)

## 3. `BURN` address
`BURN` addresses would be a new address type in CP, where the address would need to be verified as a burn address in some way, after which point, the dispenser could be created. 

# `OWNED` Address Validation
I am thinking a signed message included in the tx is the best method... also think if we standardized the message signed, we could get away with only passing the message signature in the tx, not the message and message signture.

# `BURN` Address Validation
While there is no perfect method to verify a burn address, we could probably come up with some standarized format which would support most burn addresses. Perhaps a rule of "must contain a certain number of X characters before address signature" (ie. 1CounterpartyXXXXXXXXXXXXXXXUWLpVr).

# Summary
The general idea here is :
1. Stop people from opening dispensers on addresses they don't own (except burn addresses)
2. Make it easier to operate/manage dispensers remotely (manage `OWNED` addresses from `SOURCE` address)

Thoughts?

**Comment by loon3** [comment]
I love the idea of being able to close dispensers from the source address.  This seems like it would be a relatively straightforward validation check which would allow for much smoother dispenser operation and wallet integrations.  I also can't think of any reason not to allow it.

Regarding OWNED, I don't think this is necessary.  If Bob were to open a dispenser on Alice's address, then Alice could simply close the dispenser and receive all the assets inside (since she controls the address).  We talked about having this sort of limitation in the past but it doesn't necessarily have any additional spam/harassment concerns then simply sending assets to people that may not want them.

Regarding BURN, this is a very gray area and even though Counterparty utilized a burn address in its inception its something we should probably get away from as we did by re-enabling the Destroy function.  What we could do is have a dispenser that is activated by a new dispenser message type that just contains the message id and the txid of the dispenser (similar to a cancel order ) and the burn amount is added as a value to the OP_RETURN output.  Since OP_RETURN outputs cannot be spent, this is the only way to truly burn BTC.

**Reply by loon3** [reply]
In my previous comment I confused "source" address with the originating address.  

Here's a better explanation of what I mean... 

When a valid dispenser open tx is logged in the dispensers table, a parameter ```origin``` is added to the entry where```origin = tx['source']```.  This is the address used to create the transaction. 

Example of updated ```get_dispensers``` api request...

        {
            "tx_index": 1468609,
            "tx_hash": "4a960783ac0594d25da3d65109e4109e7c3c72b7e7adc0d148f57ebc6909e4aa",
            "block_index": 600773,
            "source": "1FwkKA9cqpNRFTpVaokdRjT9Xamvebrwcu",
            "asset": "XCP",
            "give_quantity": 100000000,
            "escrow_quantity": 10000000000,
            "satoshirate": 75000,
            "status": 10,
            "give_remaining": 0,
            "oracle_address": null, 
            "origin": "1FwkKA9cqpNRFTpVaokdRjT9Xamvebrwcu"
        }
        
If the dispenser is opened on an empty address, ```origin``` will be different than ```source``` (also referred to as the dispenser address).  Otherwise, the ```origin``` and ```source``` will be the same.  The purpose of ```origin``` is to facilitate the validation check to allow for ```origin``` to close or refill any dispenser that it opens on an empty address.  

When refilling or closing a dispenser from the an origin address, we can use the existing ```open_address``` parameter in the API to specify the source address for the dispenser.  Upon closing the dispenser, the assets are credited to the origin address.

There's no API changes required to add origin addresses.

**Comment by ghost** [comment]
Dispenser Improvements： more support for physical goods than Virtual digital goods like NFT/Token only :


100% agree with the ideas more easy control of dispensers over the source addr

I prefer to think the priority right now is  more for implementing a feature that allows clients to easily send MSG (like to their address  size. email etc., )to the dispenser deployer when purchasing items like Pepecash especially physical commodities, even some chat features .

there is huge demand for this feature like Cross-border e-commerce input fields when you check out.

**Comment by Jpja** [comment]
Is the underlying issue that dispensers are being opened on active Bitcoin addresses not related to the seller? Dispenses are triggered every time btc is received without the buyer neither aware of nor intending to get those tokens. Since setting up a dispenser costs only one tx fee but can trigger many (hundreds, even thousands) of dispenses, each adding to the CP DB, this is a concern.

Can't we redefine an `EMPTY` address to have no Bitcoin history at all? (Rather than no Counterparty history, as it is now.) Won't this solve the problem and be in line with the intention of allowing dispensers on empty addresses only? Is there any technical reason not to do this?

Another solution could be to charge an XCP fee per dispense. Would work something like this; when the dispenser is set up, and if it holds >10 dispenses, then escrow XCP from the seller.

_Regarding closing dispensers from `SOURCE`. I'm in 100% in favor of this._

**Reply by jdogresorg** [reply]
1. YES, I think we should redefine `EMTPY` to mean no BTC history.

2. Joe looney has come up with a much more [elegant solution](https://gist.github.com/loon3/d103e831c26cba61ffc415085151f3a2) of simply adding an `origin` field to dispensers table, and tweaking the dispenser rules to allow `SOURCE` to refill/close a dispenser if `SOURCE`==`origin`.

I think @loon3's `origin` field addition is the best path forward for CP, and is better than this OWNED/BURNED addresses idea I proposed 👍
